File: LEETCODE/src/prob845.java
class prob845 {
    public static int longestMountain(int[] arr) {
        if (arr.length <= 2){
            return 0;
        }
        int left = 0;
        int right = 1;
        int fault = 0;
        int res = 0;
        while (right < arr.length){
            if (arr[right] <= arr[left] && right - left == 1){
                left = right;
                right ++;
                continue;
            }
            if (arr[right] == arr[right - 1]){
                fault ++;
            }
            if (right < arr.length - 1){
                if (arr[right] > arr[right - 1] && arr[right] <= arr[right + 1] || arr[right] >= arr[right - 1] && arr[right] >= arr[right + 1] || arr[right] <= arr[right - 1] && arr[right] > arr[right + 1]){
                    right ++;
                }
                else{
                    if (fault == 0){
                        int len = right - left + 1;
                        res = Math.max(len, res);
                    }
                    fault = 0;
                    left = right;
                    right ++;
                }
            }
            if (right == arr.length - 1){
                if (arr[right] >= arr[right - 1]){
                    return res;
                }
                if (fault == 0){
                    int len = right - left + 1;
                    res = Math.max(len, res);
                }
                fault = 0;
                left = right;
                right ++;
            }
        }
        return res;
    }
}

File: LEETCODE/src/prob845.cpp
class Solution {
public:
    int longestMountain(vector<int>& arr) {
        int res = 0;
        int i = 0;
        while (i < arr.size()){
            int j = i;
            if (j + 1 < arr.size() && arr[j] < arr[j + 1]){
                while (j + 1 < arr.size() && arr[j] < arr[j + 1]){
                    j ++;
                }
                if (j + 1 < arr.size() && arr[j] > arr[j + 1]){
                    while (j + 1 < arr.size() && arr[j] > arr[j + 1]){
                        j ++;
                    }
                    res = max(j - i + 1, res);
                }
            }
            i = max(i + 1, j);
        }
        return res;
    }
};
